src/logger/src/logger.rs

```diff
@@ -500,37 +451,95 @@ impl Log for Logger {
+
+    struct LogWriter {
+        buf: Arc<Mutex<Vec<u8>>>,
+    }
+
+    impl Write for LogWriter {
+        fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+            let mut data = self.buf.lock().unwrap();
+            data.append(&mut buf.to_vec());
+
+            Ok(buf.len())
+        }
+
+        fn flush(&mut self) -> std::io::Result<()> {
+            Ok(())
+        }
+    }
+
+    struct LogReader {
+        buf: Arc<Mutex<Vec<u8>>>,
+    }
+
+    impl LogReader {}
+
+    impl Read for LogReader {
+        fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
+            let mut data = self.buf.lock().unwrap();
+
+            let len = std::cmp::min(data.len(), buf.len());
+            buf[..len].copy_from_slice(&data[..len]);
 
-    fn validate_logs(log_path: &str, expected: &[(&'static str, &'static str)]) -> bool {
-        let f = File::open(log_path).unwrap();
-        let mut reader = BufReader::new(f);
-
-        let mut line = String::new();
-        // The first line should contain the app header.
-        reader.read_line(&mut line).unwrap();
-        assert!(line.contains(TEST_APP_HEADER));
-        for tuple in expected {
-            line.clear();
-            // Read an actual log line.
-            reader.read_line(&mut line).unwrap();
-            assert!(line.contains(&TEST_INSTANCE_ID));
-            assert!(line.contains(&tuple.0));
-            assert!(line.contains(&LOG_SOURCE));
-            assert!(line.contains(&tuple.1));
+            data.drain(..len);
+
+            Ok(len)
         }
-        false
+    }
+
+    fn log_channel() -> (LogWriter, LogReader) {
+        let buf = Arc::new(Mutex::new(vec![]));
+        (
+            LogWriter { buf: buf.clone() },
+            LogReader { buf: buf.clone() },
+        )
+    }
+
+    fn log(logger: &Logger, level: Level, msg: &str) {
+        logger.log(
+            &log::Record::builder()
+                .level(level)
+                .args(format_args!("{}", msg))
+                .file(Some(LOG_SOURCE))
+                .line(Some(LOG_LINE))
+                .build(),
+        );
+    }
+
+    fn validate_log(mut log_reader: Box<&mut dyn Read>, expected: &str) {
+        let mut log = Vec::<u8>::new();
+        log_reader.read_to_end(&mut log).unwrap();
+
+        assert!(log.len() >= expected.len());
+        assert_eq!(expected.as_bytes(), &log[log.len() - expected.len()..]);
+    }
+
+    fn create_logger() -> Logger {
+        let logger = Logger::new();
+        logger.set_max_level(log::LevelFilter::Info);
+        logger.set_instance_id(TEST_INSTANCE_ID.to_string());
+
+        logger
+    }
+
+    fn init_logger(logger: &Logger) -> LogReader {
+        let (writer, mut reader) = log_channel();
+        assert!(logger
+            .init(TEST_APP_HEADER.to_string(), Box::new(writer))
+            .is_ok());
+        validate_log(Box::new(&mut reader), &format!("{}\n", TEST_APP_HEADER));
+
+        reader
     }
```

Comment:
Hmm, it looks like you are using only `init_logger()` from them for testing the static logger. Do you plan on adding more tests? I think that `reader()` can be used also for the static logger, but I might be missing something.

Reply:
Yes, but if we can't use `create()` for the static logger, then we won't be able to use any of the other methods. Because we won't be able to use this new class.

Reply:
Sorry for the misunderstanding. I thought about something else. 

I tried your suggestion and I'm not sure about it. I find this code more confusing after adding test methods in the `Logger` class.  For example having a method like `log_data()` inside `Logger` is a bit strange and it would make me confused if I looked at these unit tests for the first time. Personally I would keep things simple for the moment

Reply:
It's a common practice in our repo to extend a struct functionality in tests where this is possible. Some examples would be: `devices/src/virtio/net/device.rs`, `dumbo/src/ns.rs` and `vmm/src/device_manager/mmio.rs`. For consistency, I think we should stick to this. We also avoid this way orphan functions (even if they are tests functions).

I also believe that, if we use the right naming for those functions, the unit tests won't be confusing. For me it is actually less confusing with the `impl Logger` approach because it becomes clearer why and for what type of object (in this case the Logger) are those functions implemented; you will also reduce significantly `logger` word occurrences in function names and parameters. This approach also makes it easier to add future functions, for example the one that takes the 3 boolean parameters I suggested in this PR in another thread.

Reply:
Ok. Done. Please take a look !